rust/src/write/writer.rs

```diff
@@ -0,0 +1,606 @@
+            .set_compression(Compression::SNAPPY)
+            .build();
+
+        Ok(Self {
+            storage,
+            table_uri: table_uri.clone(),
+            arrow_schema_ref: schema,
+            writer_properties,
+            partition_columns: partition_columns.unwrap_or_default(),
+            arrow_writers: HashMap::new(),
+        })
+    }
+
+    /// Creates a DeltaWriter to write to the given table
+    pub fn for_table(
+        table: &DeltaTable,
+        options: HashMap<String, String>,
+    ) -> Result<DeltaWriter, DeltaWriterError> {
+        let storage = get_backend_for_uri_with_options(&table.table_uri, options)?;
+
+        // Initialize an arrow schema ref from the delta table schema
+        let metadata = table.get_metadata().unwrap();
+        let arrow_schema = <ArrowSchema as TryFrom<&Schema>>::try_from(&metadata.schema.clone())?;
+        let arrow_schema_ref = Arc::new(arrow_schema);
+        let partition_columns = metadata.partition_columns.clone();
+
+        // Initialize writer properties for the underlying arrow writer
+        let writer_properties = WriterProperties::builder()
+            // NOTE: Consider extracting config for writer properties and setting more than just compression
+            .set_compression(Compression::SNAPPY)
+            .build();
+
+        Ok(Self {
+            storage,
+            table_uri: table.table_uri.clone(),
+            arrow_schema_ref,
+            writer_properties,
+            partition_columns,
+            arrow_writers: HashMap::new(),
+        })
+    }
+
+    /// Divides a single record batch into into multiple according to table partitioning.
+    /// Values are written to arrow buffers, to collect data until it should be written to disk.
+    pub fn write(&mut self, values: &RecordBatch) -> Result<(), DeltaWriterError> {
+        let arrow_schema = self.partition_arrow_schema();
+        for result in self.divide_by_partition_values(values)? {
+            let partition_key =
+                DeltaWriter::get_partition_key(&self.partition_columns, &result.partition_values)?;
+            match self.arrow_writers.get_mut(&partition_key) {
+                Some(writer) => {
+                    writer.write(&result.record_batch)?;
+                }
+                None => {
+                    let mut writer = PartitionWriter::new(
+                        arrow_schema.clone(),
+                        result.partition_values,
+                        self.writer_properties.clone(),
+                    )?;
+                    writer.write(&result.record_batch)?;
+                    let _ = self.arrow_writers.insert(partition_key, writer);
+                }
+            }
+        }
+
+        Ok(())
+    }
+
+    /// Writes the existing parquet bytes to storage and resets internal state to handle another file.
+    pub async fn flush(&mut self) -> Result<Vec<Add>, DeltaWriterError> {
```

Comment:
the semantic of flush here is different from the one in the existing json writer, is there any particular reason why you to keep it different?

Reply:
Not 100% sure which part you are referring to, but there are some things i considered :). The implementation with the json writer already creates a commit. Since we are hoping to eventually use this writer in a distributed context i though we should avoid doing so here. When its referring to the error type, these should probably be harmonized. In general it may be worth exploring to have a shared writer trait that different writers could implement?

Reply:
Yes, I was referring to the flush method from the json writer: https://github.com/delta-io/delta-rs/blob/346f51a277c23be09eca373cd6d775b745d59c05/rust/src/writer.rs#L94.

> Since we are hoping to eventually use this writer in a distributed context i though we should avoid doing so here.

I think this makes sense, if so, then it's better to use a different name other than flush to avoid confusions.

> In general it may be worth exploring to have a shared writer trait that different writers could implement?

I haven't put much thought into this yet, maybe think about what are the common interfaces we want to share between writers? From a quick glance, the writer method might not be a good fit because it takes different type of input depending on the input format that needs to be supported.